Add HostProcess Container Configuration for k8s

jsturtevant opened the pull request:
Kubernetes now has support for [HostProcess Containers](https://kubernetes.io/blog/2021/08/16/windows-hostprocess-containers/) so this is a continuation of https://github.com/prometheus-community/windows_exporter/pull/581 using HostProcess containers

Requirements:

 - Kuberentes 1.22+
 - containerd 1.6 Beta+
 - WindowsHostProcessContainers feature-gate (currently in alpha in 1.22 going beta in 1.23) turned on for kube-apiserver and kubelet if using Kubernetes 1.22

jsturtevant commented:
I know the requirements are hard to come by so for anyone that wants to try this out there are few options available currently: 
- https://capz.sigs.k8s.io/topics/windows.html#using-containerd-for-windows-clusters 
- https://github.com/kubernetes-sigs/sig-windows-dev-tools

marosset commented on kubernetes/kubernetes.md:
```diff
@@ -0,0 +1,74 @@
+- containerd 1.6 Beta+
+- WindowsHostProcessContainers feature-gate (currently in alpha in 1.22) turned on for kube-apiserver and kubelet if using Kubernetes 1.22
+
+> IMPORTANT: This does not work unless you are specifically targeting Host Process Containers with Containerd (Docker doesn't have support).  The image will build and run but will be **not** able to access the host.
```
Should we add an init container to the ds-yaml below that runs a check to make sure everything is configured properly to run host-process containers?
Something as simple as https://github.com/kubernetes/kubernetes/blob/6d9008b1b086008caed5a9d8ef150cca0abaefa9/test/e2e/windows/host_process.go#L109 should be sufficient.
I think this could prevent a lot of confusion / time spent debugging.

marosset replied:
I guess you'd have to this too and then make sure %COMPUTERNAME% in the container matches the node name env var 
https://github.com/kubernetes/kubernetes/blob/6d9008b1b086008caed5a9d8ef150cca0abaefa9/test/e2e/windows/host_process.go#L505-L513

perithompson replied:
+1 for the daemon set yaml but wonder if, once you're on k8s then you know if its hostprocess or not anyway?  good one to think about.  Could this be done in the binary to check it has access to the requested counters at startup?

perithompson replied:
or output running as a user in the log message?

jsturtevant replied:
I like the idea of checking the current user and giving a warning in logs. 

I like the idea of adding an init container to the docs for hostprocess as a debugging tool but having across the board for projects seems a bit burdensome on the end user.

cc: @brasmith-ms on possible troubleshooting tips for hostprocess docs

jsturtevant replied:
I found that with nanoserver, the container will not start due to https://github.com/prometheus-community/windows_exporter/blob/f84f54afdaa215259b443f2d6011c99902a181dd/exporter.go#L250 not having required libraries in nano.  This means if someone didn't add the `hostprocess` fields to the spec it would fail to start the container if using the image produced with the dockerfile here.  

Maybe we want to add a warning here too for usability?

With these updates, If someone built there own image and used servercore the container will start and get the following:

```
k logs windows-exporter-hbhvm                        
time="2021-12-01T21:07:39Z" level=warning msg="No filters specified for process collector. This will generate a very large number of metrics!" source="process.go:58"
time="2021-12-01T21:07:39Z" level=info msg="Running as User Manager\\ContainerAdministrator" source="exporter.go:354"
time="2021-12-01T21:07:39Z" level=warning msg="Running as a preconfigured Windows Container user. This may mean you do not have Windows HostProcess containers configured correctly and some functionality will not work as expected." source="exporter.go:356"
time="2021-12-01T21:07:39Z" level=info msg="Enabled collectors: process" source="exporter.go:359"
time="2021-12-01T21:07:39Z" level=info msg="Starting windows_exporter (version=latest, branch=use-hostprocess, revision=95e15738886f4ddd40cc277ae1821a0b0c6f5c0a)" source="exporter.go:411"
time="2021-12-01T21:07:39Z" level=info msg="Build context (go=go1.17.3, user=root@buildkitsandbox, date=20211201-20:43:54)" source="exporter.go:412"
time="2021-12-01T21:07:39Z" level=info msg="Starting server on :9182" source="exporter.go:415"
time="2021-12-01T21:07:39Z" level=info msg="TLS is disabled." source="gokit_adapter.go:38"

```

marosset commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+# Note this image doesn't really matter for hostprocess
```
I think the biggest issue w/ using a mis-matched image for host-process containers is added overhead from needing to pull the nanoserver base layers for a different OS version.
This will work but long-term maybe we should at least have a ltsc2022 image too.

jsturtevant replied:
added logic to build both 2019 and 2022

perithompson commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+FROM $BASE
+
+ENV PATH="C:\Windows\system32;C:\Windows;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;"
```
```suggestion
ENV PATH="C:\Windows\system32;C:\Windows;"
```

perithompson commented on kubernetes/windows-exporter-daemonset.yaml:
```diff
@@ -0,0 +1,58 @@
+      - args: 
+        - --config.file=%CONTAINER_SANDBOX_MOUNT_POINT%/config.yml
+        name: windows-node-exporter
+        image: jsturtevant/windows-exporter:latest
```
could we get this in the same place as node-exporter?

jsturtevant replied:
yes, I need some guidance from @windows_exporter-reviewers on naming and how to get the image published. I took some of the constructs form node_exporter but not sure if tooling is in place to push the images or if I have the correct docker repository

jsturtevant replied:
@carlpett @breed808 do you know what needs to be done here?

jsturtevant replied:
just as an fyi - we don't need windows build machines for this image. We can use buildx running on Linux to build the images. The requirement is that you must have a registry to export too when building since we must use `--output=type=registry` with buildx.

perithompson commented on kubernetes/windows-exporter-daemonset.yaml:
```diff
@@ -0,0 +1,58 @@
+      enabled: cpu,cs,net,logical_disk,os,system,service,container
+    collector:
+      service:
+        services-where: "Name='windows_exporter'"
```
this service wont exist if running in a container?

perithompson replied:
we could user containerd and kubelet as an example?

jsturtevant replied:
good catch :-)

jsturtevant commented on kubernetes/windows-exporter-daemonset.yaml:
```diff
@@ -0,0 +1,58 @@
+data:
+  config.yml: |
+    collectors:
+      enabled: cpu,cs,net,logical_disk,os,system,service,container
```
not in particular. I personally tend to try to be explicit but for the quick-start this makes sense

jsturtevant commented on kubernetes/windows-exporter-servicemonitor.yaml:
```diff
@@ -0,0 +1,16 @@
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
```
Mimicked the service monitor from kube-prometheus.  I will look into this, I am not familiar with it

jsturtevant replied:
updated to use podmonitor.  This simplified the different parts. Thanks for sharing!

jsturtevant commented on kubernetes/kubernetes.md:
```diff
@@ -0,0 +1,74 @@
+
+## Kubernetes Resources
+
+Before beginning you need to deploy the [prometheus operator](https://github.com/prometheus-operator/prometheus-operator) to your cluster. You can use a project like https://github.com/prometheus-operator/kube-prometheus to get started.
```
updated the wording

jsturtevant commented on kubernetes/windows-exporter-daemonset.yaml:
```diff
@@ -0,0 +1,58 @@
+kind: DaemonSet
+metadata:
+  labels:
+    app: windows-node-exporter
```
updated

jsturtevant commented:
@carlpett I've addressed all the feedback. Ready for a final review and to sort out the image pushing mechanism.  We only need Linux machine and Container Registry since we are using buildx to create the image.

carlpett reviewed (approved):
All looks good to me, thanks a lot!
We still need the credentials, though, and looks like SuperQ is busy. Perhaps @RichiH has an idea on who could help out?

jsturtevant commented:
Let me know if can help in anyway.

jsturtevant commented:
@carlpett anything we can do to move this along? Is there a community meeting I can add it to the agenda? or maybe start a conversation in slack?

carlpett commented:
Hi @jsturtevant,
We actually just completed a migration from AppVeyor to Github Actions, which means we should now be able to utilize the builtin container registry here without account management headaches.

So I think the next step would be amending the release step to build and push the image. Something you'd be comfortable doing?

jsturtevant commented:
sweet! I'll take a look tomorrow and see if I have questions. Thanks!

jsturtevant commented:
Looked into it a bit and looks like the github action jobs are running on Windows.  Buildx doesn't work on windows so I will need to adjust this a bit to build containers on  Windows.  Shouldn't be a big deal.

I was also building the image in the container but you have the image on the build machine so I could just copy that into the container.  This way its the same executable that is being released.

breed808 commented:
It's worth noting that the OS can be set at the job level, so in theory you could add another job running Linux and use the [artifacts](https://docs.github.com/en/actions/advanced-guides/storing-workflow-data-as-artifacts) from the [`Build`](https://github.com/prometheus-community/windows_exporter/blob/a56ec9166bd7408a0ab5005b14a9ca032664c60d/.github/workflows/ci.yml#L114) job.

jsturtevant commented:
> It's worth noting that the OS can be set at the job level, so in theory you could add another job running Linux and use the [artifacts](https://docs.github.com/en/actions/advanced-guides/storing-workflow-data-as-artifacts) from the [`Build`](https://github.com/prometheus-community/windows_exporter/blob/a56ec9166bd7408a0ab5005b14a9ca032664c60d/.github/workflows/ci.yml#L114) job.

I think it makes sense to have the build work on Windows since rest of the tooling is working there.

I've pushed the changes to make this work on Windows github action runner as separate commit.  I made minimal changes to make it work but let me know if you want a different approach.  Can squash if you want me too.

Note to build the multi arch image needed to use the WS2022 runners, which apparently don't have the `gitversion` tool installed by default.

carlpett commented:
This looks good to me, great work! :clap:
Did you have the possibility to test it out end-to-end somewhere? Otherwise I think the best we can do is merge, make a draft release and see what happens?

jsturtevant commented:
I built an image with this code here: https://hub.docker.com/repository/docker/jsturtevant/windows-exporter if you want to try it out (requires k8s 1.22+ and containerd 1.6+)

I will run a quick validation this afternoon.

jsturtevant commented:
I validated it is working with that image. I will double check once there is a released image

```
k get nodes -o wide                                                                                                                                                                                                                                                 
NAME                                       STATUS   ROLES                  AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                         KERNEL-VERSION     CONTAINER-RUNTIME                                                                                                                                                                                        
win-conta-qzl6z                            Ready    <none>                 32m   v1.22.2   10.1.0.5      <none>        Windows Server 2019 Datacenter   10.0.17763.2237    containerd://1.6.0-beta.0                                                                                                                                                                                
win-conta-snwfp                            Ready    <none>                 32m   v1.22.2   10.1.0.7      <none>        Windows Server 2019 Datacenter   10.0.17763.2237    containerd://1.6.0-beta.0                                                                                                                                                                                
win-containerd-32766-control-plane-9lltd   Ready    control-plane,master   35m   v1.22.2   10.0.0.4      <none>        Ubuntu 20.04.3 LTS               5.8.0-1041-azure   containerd://1.5.5                                                                                                                                                                                       
win-containerd-32766-md-0-fclv9            Ready    <none>                 33m   v1.22.2   10.1.0.4      <none>        Ubuntu 20.04.3 LTS               5.8.0-1041-azure   containerd://1.5.5                                                                                                                                                                                       
win-containerd-32766-md-0-m4zhn            Ready    <none>                 33m   v1.22.2   10.1.0.6      <none>        Ubuntu 20.04.3 LTS               5.8.0-1041-azure   containerd://1.5.5

                                                                                                                                                                                       
[win-containerd-32766-admin@win-containerd-32766|monitoring]  ~/projects/windows_exporter   use-hostprocess  kgp                                                                                                                                                                                                                                                                 
NAME                                  READY   STATUS    RESTARTS   AGE    IP                NODE                                       NOMINATED NODE   READINESS GATES                                                                                                                                                                                                             
alertmanager-main-0                   2/2     Running   0          21m    192.168.24.130    win-containerd-32766-md-0-fclv9            <none>           <none>                                                                                                                                                                                                                      
alertmanager-main-1                   2/2     Running   0          21m    192.168.178.194   win-containerd-32766-md-0-m4zhn            <none>           <none>                                                                                                                                                                                                                      
alertmanager-main-2                   2/2     Running   0          21m    192.168.24.129    win-containerd-32766-md-0-fclv9            <none>           <none>                                                                                                                                                                                                                      
blackbox-exporter-6798fb5bb4-phvlt    3/3     Running   0          21m    192.168.24.131    win-containerd-32766-md-0-fclv9            <none>           <none>                                                                                                                                                                                                                      
grafana-767fcb6796-tgds5              1/1     Running   0          21m    192.168.178.195   win-containerd-32766-md-0-m4zhn            <none>           <none>                                                                                                                                                                                                                      
kube-state-metrics-6c699dfb8-c8r45    3/3     Running   0          21m    192.168.24.132    win-containerd-32766-md-0-fclv9            <none>           <none>                                                                                                                                                                                                                      
node-exporter-bcgmj                   2/2     Running   0          21m    10.1.0.6          win-containerd-32766-md-0-m4zhn            <none>           <none>                                                                                                                                                                                                                      
node-exporter-hpm78                   2/2     Running   0          21m    10.0.0.4          win-containerd-32766-control-plane-9lltd   <none>           <none>                                                                                                                                                                                                                      
node-exporter-sk9vf                   2/2     Running   0          21m    10.1.0.4          win-containerd-32766-md-0-fclv9            <none>           <none>                                                                                                                                                                                                                      
prometheus-adapter-7dc46dd46d-twgwp   1/1     Running   0          21m    192.168.178.196   win-containerd-32766-md-0-m4zhn            <none>           <none>                                                                                                                                                                                                                      
prometheus-adapter-7dc46dd46d-zh86k   1/1     Running   0          21m    192.168.24.133    win-containerd-32766-md-0-fclv9            <none>           <none>                                                                                                                                                                                                                      
prometheus-k8s-0                      2/2     Running   0          21m    192.168.24.134    win-containerd-32766-md-0-fclv9            <none>           <none>                                                                                                                                                                                                                      
prometheus-k8s-1                      2/2     Running   0          21m    192.168.178.197   win-containerd-32766-md-0-m4zhn            <none>           <none>                                                                                                                                                                                                                      
prometheus-operator-5c875b748-h98gq   2/2     Running   0          22m    192.168.178.193   win-containerd-32766-md-0-m4zhn            <none>           <none>                                                                                                                                                                                                                      
windows-exporter-5c5l7                1/1     Running   0          91s    10.1.0.7          win-conta-snwfp                            <none>           <none>                                                                                                                                                                                                                      
windows-exporter-drjp9                1/1     Running   0          107s   10.1.0.5          win-conta-qzl6z                            <none>           <none>    

                                                                                                                                                                                                                  
[win-containerd-32766-admin@win-containerd-32766|monitoring]  ~/projects/windows_exporter   use-hostprocess  kubectl get --raw=/api/v1/namespaces/monitoring/pods/windows-exporter-5c5l7/proxy/metrics                                                                                                                                                                           
# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.                                                                                                                                                                                                                                                                                         
# TYPE go_gc_duration_seconds summary                                                                                                                                                                                                                                                                                                                                               
go_gc_duration_seconds{quantile="0"} 0                                                                                                                                                                                                                                                                                                                                              
go_gc_duration_seconds{quantile="0.25"} 0                                                                                                                                                                                                                                                                                                                                           
go_gc_duration_seconds{quantile="0.5"} 0                                                                                                                                                                                                                                                                                                                                            
go_gc_duration_seconds{quantile="0.75"} 0                                                                                                                                                                                                                                                                                                                                           
go_gc_duration_seconds{quantile="1"} 0.0012611                                                                                                                                                                                                                                                                                                                                      
go_gc_duration_seconds_sum 0.0029943                                                                                                                                                                                                                                                                                                                                                
go_gc_duration_seconds_count 21                                                                                                                                                                                                                                                                                                                                                     
# HELP go_goroutines Number of goroutines that currently exist.                                                                                                                                                                                                                                                                                                                     
# TYPE go_goroutines gauge                                                                                                                                                                                                                                                                                                                                                          
go_goroutines 9                                                                                                                                                                                                                                                                                                                                                                     
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
....
```

jsturtevant commented:
do you want me to squash the commits?

jsturtevant commented on kubernetes/windows-exporter-daemonset.yaml:
```diff
@@ -0,0 +1,61 @@
+      - args: 
+        - --config.file=%CONTAINER_SANDBOX_MOUNT_POINT%/config.yml
+        name: windows-exporter
+        image: jsturtevant/windows-exporter:latest
```
@carlpett  I need to update this but currently not pushing a latest tag.  We won't have a tag until the git repository is tagged and released.  Thoughts?

the current format is `ghcr.io/prometheus-community/windows-exporter:<versionnumber>`

carlpett replied:
Based on other unreleased changes, the next version is going to be 0.18.
I'm not sure if we want a `latest` tag? On the other hand, documenting it as a fixed version, I doubt we'll remember to update it...

carlpett replied:
Thinking some more, I think we'll want a "bleeding edge" image. For example, the `node_exporter` has a `master` tag. They also have a `latest`, pointing to latest release. I'm personally not a fan, but I suppose there's probably high demand for it?

jsturtevant replied:
added a "bleeding" edge image.  let me know if you want me to squash the commits

jsturtevant commented:
e2e test failed, It might have been a flake? I don't seem to be able to re-triger the job

jsturtevant commented:
> e2e test failed, It might have been a flake? I don't seem to be able to re-triger the job

`git commit --amend --no-edit -s` and `git push -force` worked

jsturtevant commented:
must be something else. I don't see anything in the directly in the output

jsturtevant commented:
@carlpett @breed808 Looking at a few other open prs which have this error and this looks like it might be broken on the main branch?

breed808 commented:
Yes, you'll need to rebase this PR on `master` to pick up the CI fix. See #924 for context.

breed808 commented:
@jsturtevant are you happy for this to be merged? I can't see any open threads besides the bleeding edge image, which has been implemented.

jsturtevant commented:
Yes let's do it 🥳 I just rebased to clean up the commits.
